hangman/lib/hangman/impl.ex

```diff
@@ -0,0 +1,95 @@
+defmodule Hangman.Impl do
+  def new_game() do
+    %Hangman.Game{%Hangman.Game{} | letters: Dictionary.random_word |> String.codepoints}
+  end
+
+  def tally(game = %Hangman.Game{}) do
+    %{
+      game_state: game.game_state,
+      turns_left: game.turns_left,
+      letters: get_tally_letters(game),
+      used: game.used,
+      last_guess: game.last_guess
+    }
+  end
+
+  def make_move(game = %Hangman.Game{}, guess) when byte_size(guess) == 1 do
+    updated_game = %Hangman.Game{game | last_guess: guess} |>
+                   check_guess() |>
+                   update_used_letters() |>
+                   update_turns_left |>
+                   check_game_over
+
+    
+    #game = %Hangman.Game{game | last_guessed: guess, turns_left: game.turns_left - 1}
+    {updated_game, tally(updated_game)}
+  end
+
+  def make_move(_game, _guess) do
+    {:error, "guess should be one letter"}
+  end
+
+  def get_tally_letters(%Hangman.Game{game_state: :lost, letters: letters}) do
+    letters
+  end
+  
+  def get_tally_letters(%Hangman.Game{letters: letters, used: used}) do
+    letters
+    |> hide_chars(used)
+  end
+
+  def hide_chars([head| tail], used_letters) do
+    if (Enum.member?(used_letters, head)) do
+        [head, hide_chars(tail, used_letters)]
+    else
+        ["_", hide_chars(tail, used_letters)]
+    end
+  end
+
+  def hide_chars([], _used_letters) do
+    []
+  end
+
+  def check_guess(game = %Hangman.Game{used: used, letters: letters, last_guess: guess}) do
+    cond do
```

Comment:
Normally I'm taking a couple of points off for conditional logic.

However, in this case, it reads really well, and I'm letting it pass.

However, as you clearly get what you're doing, I am going to nit pick.

With a `cond`, your writing a series of positive assertions and corresponding actions. "when _this_ do _that_".

The first two heads in your cond follow this pattern. But the last one doesn't, because it doesn't really mean what it says.

You have

~~~ elixir
guess not in letters ->
  %Hangman.Game{game | game_state: :bad_guess}
~~~

but that's not what you mean. At this point it is nothing to do with the guess. This is the catchall clause, so it should be written

~~~ elixir
true ->
  %Hangman.Game{game | game_state: :bad_guess}
~~~